src/query/storage/block.go

```diff
@@ -31,37 +31,81 @@ import (
 
 // FetchResultToBlockResult converts a fetch result into coordinator blocks
 func FetchResultToBlockResult(result *FetchResult, query *FetchQuery) (block.Result, error) {
-	alignedSeriesList, err := result.SeriesList.Align(query.Start, query.End, query.Interval)
-	if err != nil {
-		return block.Result{}, err
-	}
-
-	multiBlock, err := newMultiSeriesBlock(alignedSeriesList, query)
+	multiBlock, err := NewMultiSeriesBlock(result.SeriesList, query)
 	if err != nil {
 		return block.Result{}, err
 	}
 
 	return block.Result{
-		Blocks: []block.Block{multiBlock},
+		Blocks: []block.Block{NewMultiBlockWrapper(multiBlock)},
 	}, nil
 }
 
+// NewMultiBlockWrapper returns a block wrapper over an unconsolidated block
+func NewMultiBlockWrapper(unconsolidatedBlock block.UnconsolidatedBlock) block.Block {
+	return &multiBlockWrapper{
+		unconsolidated: unconsolidatedBlock,
+	}
+}
+
+type multiBlockWrapper struct {
+	unconsolidated block.UnconsolidatedBlock
+	consolidated   block.Block
+	mu             sync.Mutex
+}
+
+func (m *multiBlockWrapper) Unconsolidated() (block.UnconsolidatedBlock, error) {
+	return m.unconsolidated, nil
+}
+
+func (m *multiBlockWrapper) StepIter() (block.StepIter, error) {
+	if err := m.tryConsolidate(); err != nil {
+		return nil, err
+	}
+
+	return m.consolidated.StepIter()
+}
+
+func (m *multiBlockWrapper) tryConsolidate() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.consolidated == nil {
+		consolidated, err := m.unconsolidated.Consolidate()
+		if err != nil {
+			return err
+		}
+
+		m.consolidated = consolidated
+	}
+
+	return nil
+}
+
+func (m *multiBlockWrapper) SeriesIter() (block.SeriesIter, error) {
+	if err := m.tryConsolidate(); err != nil {
+		return nil, err
+	}
+
+	return m.consolidated.SeriesIter()
+}
+
+func (m *multiBlockWrapper) Close() error {
+	return m.unconsolidated.Close()
```

Comment:
Should also close consolidated?

Reply:
the person using that block is supposed to close ?

Reply:
Can you add a comment indicating that it should only close unconsolidated in case someone accidentally notices it and tries to close the consolidated here